### PR TITLE
Access collision in `Context` dictionary

### PR DESCRIPTION
**Pull request by colibrishin**
The access collision was observed in linux-unity-test from `ConsensusContext<T>._contexts` dictionary after https://github.com/planetarium/libplanet/pull/2390 patch.

### CODE REVIEW

**Comment by greymistcube** [comment]
Just my two cents, but as `ConsensusContext` is accessed asynchronously and observing how `_context` is used, I don't think simply using a `ConcurrentDictionary` is a proper solution. Either all dictionary access method should be rewritten with concurrency in mind if we are to use `ConcurrentDictionary` **or** use a `lock` on `_context`. 🙄

**Comment by limebell** [comment]
How about using a lock instead of `ConcurrentDictionary`? The logic of `ConsensusContext` is not considered to be used concurrently IMO.

**Comment by colibrishin** [comment]
Changed from `ConcurrentDictionary` to `lock` and set it to where race conditions occurred.

File: Libplanet.Net/Consensus/ConsensusContext.cs [hunk 1]
```diff
@@ -324,7 +333,7 @@ private void ClearOldLastCommitCache(long maxSize = 30)
         }
 
         /// <summary>
-        /// Discard and remove all contexts that has lower or equal height with
+        /// Discard and remove all contexts that has lower height with
```
Comment (limebell) [inline Libplanet.Net/Consensus/ConsensusContext.cs:336]: 👍

**Review by limebell — APPROVED** [review]

File: Libplanet.Net/Consensus/ConsensusContext.cs [hunk 2]
```diff
@@ -199,14 +202,17 @@ public void NewHeight(long height)
...
             {
-                _contexts[height] = new Context<T>(
-                    this,
-                    _blockChain,
-                    height,
-                    _privateKey,
-                    _getValidators(height).ToList(),
-                    Step.Default,
-                    contextTimeoutOptions: _contextTimeoutOption);
+                lock (_contextLock)
```
Comment (greymistcube) [inline Libplanet.Net/Consensus/ConsensusContext.cs:205]: This should be moved one layer up I think?

File: Libplanet.Net/Consensus/ConsensusContext.cs [hunk 3]
```diff
@@ -251,13 +257,16 @@ public void HandleMessage(ConsensusMessage consensusMessage)
...
             if (!_contexts.ContainsKey(height))
             {
-                _contexts[height] = new Context<T>(
-                    this,
-                    _blockChain,
-                    height,
-                    _privateKey,
-                    _getValidators(height).ToList(),
-                    _contextTimeoutOption);
+                lock (_contextLock)
```
Comment (greymistcube) [inline Libplanet.Net/Consensus/ConsensusContext.cs:260]: Same as above

File: Libplanet.Net/Consensus/ConsensusContext.cs (hunk 2 above)
Comment (colibrishin) [reply Libplanet.Net/Consensus/ConsensusContext.cs]: https://github.com/planetarium/libplanet/blob/ab82ef9d1219224f7d877fb27b45004a6e7e1e36/Libplanet.Net/Consensus/ConsensusContext.cs#L206-L223

Amended.

File: Libplanet.Net/Consensus/ConsensusContext.cs (hunk 3 above)
Comment (colibrishin) [reply Libplanet.Net/Consensus/ConsensusContext.cs]: https://github.com/planetarium/libplanet/blob/ab82ef9d1219224f7d877fb27b45004a6e7e1e36/Libplanet.Net/Consensus/ConsensusContext.cs#L261-L275

Amended.

**Pull request merged**